File: backend/app/api/routes/roles.py
from fastapi import APIRouter, Body, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, require_cap
from app.core.capabilities import CAP_KEYS, capabilities_grouped, user_capabilities
from app.core.database import get_db
from app.models import Role, User, UserRole

router = APIRouter(tags=["roles"])


def _role_out(r: Role) -> dict:
    return {
        "id": r.id, "key": r.key, "name": r.name,
        "is_system": r.is_system, "is_superadmin": r.is_superadmin, "is_default": r.is_default,
        "capabilities": list(r.capabilities or []),
    }


@router.get("/me/capabilities")
def my_capabilities(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    """Права и роли текущего пользователя — для навигации/гейтинга на фронте."""
    from app.core.capabilities import user_roles as _ur
    return {
        "capabilities": sorted(user_capabilities(db, user)),
        "roles": [r.key for r in _ur(db, user)],
    }


@router.get("/capabilities")
def list_capabilities(_: User = Depends(require_cap("roles.manage"))):
    """Каталог всех прав (сгруппированный) — для редактора ролей."""
    return capabilities_grouped()


@router.get("/roles")
def list_roles(db: Session = Depends(get_db), _: User = Depends(require_cap("roles.manage"))):
    return [_role_out(r) for r in db.query(Role).order_by(Role.is_system.desc(), Role.id).all()]


@router.post("/roles", status_code=status.HTTP_201_CREATED)
def create_role(payload: dict = Body(...), db: Session = Depends(get_db), _: User = Depends(require_cap("roles.manage"))):
    key = (payload.get("key") or "").strip().lower().replace(" ", "_")
    name = (payload.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=400, detail="Укажите название роли")
    if not key:
        key = name.lower().replace(" ", "_")
    if db.query(Role).filter(Role.key == key).first():
        raise HTTPException(status_code=400, detail="Роль с таким ключом уже существует")
    caps = [c for c in (payload.get("capabilities") or []) if c in CAP_KEYS]
    role = Role(key=key, name=name, is_system=False, is_superadmin=False,
                is_default=bool(payload.get("is_default")), capabilities=caps)
    if role.is_default:
        db.query(Role).update({Role.is_default: False})
    db.add(role)
    db.commit()
    db.refresh(role)
    return _role_out(role)


@router.put("/roles/{role_id}")
def update_role(role_id: int, payload: dict = Body(...), db: Session = Depends(get_db),
                _: User = Depends(require_cap("roles.manage"))):
    role = db.get(Role, role_id)
    if not role:
        raise HTTPException(status_code=404, detail="Роль не найдена")
    if role.is_superadmin:
        raise HTTPException(status_code=400, detail="Роль гуру не редактируется")
    if "name" in payload and payload["name"].strip():
        role.name = payload["name"].strip()
    if "capabilities" in payload:
        role.capabilities = [c for c in payload["capabilities"] if c in CAP_KEYS]
    if "is_default" in payload:
        role.is_default = bool(payload["is_default"])
        if role.is_default:
            db.query(Role).filter(Role.id != role.id).update({Role.is_default: False})
    db.commit()
    db.refresh(role)
    return _role_out(role)


@router.delete("/roles/{role_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_role(role_id: int, db: Session = Depends(get_db), _: User = Depends(require_cap("roles.manage"))):
    role = db.get(Role, role_id)
    if not role:
        raise HTTPException(status_code=404, detail="Роль не найдена")
    if role.is_system:
        raise HTTPException(status_code=400, detail="Системную роль удалить нельзя")
    db.query(UserRole).filter(UserRole.role_id == role.id).delete()
    db.delete(role)
    db.commit()


@router.get("/users/{user_id}/roles")
def get_user_roles(user_id: int, db: Session = Depends(get_db), _: User = Depends(require_cap("users.manage"))):
    ids = [ur.role_id for ur in db.query(UserRole).filter(UserRole.user_id == user_id).all()]
    return {"role_ids": ids}


@router.put("/users/{user_id}/roles")
def set_user_roles(user_id: int, payload: dict = Body(...), db: Session = Depends(get_db),
                   _: User = Depends(require_cap("users.manage"))):
    user = db.get(User, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="Пользователь не найден")
    role_ids = payload.get("role_ids") or []
    valid = {r.id for r in db.query(Role).filter(Role.id.in_(role_ids)).all()} if role_ids else set()
    db.query(UserRole).filter(UserRole.user_id == user_id).delete()
    for rid in valid:
        db.add(UserRole(user_id=user_id, role_id=rid))
    db.commit()
    return {"role_ids": sorted(valid)}

File: backend/internal/web/roles.go
package web

import (
	"net/http"
	"sort"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"manibandha/internal/caps"
	"manibandha/internal/models"
)

func roleOut(r *models.Role) map[string]any {
	c := r.Capabilities
	if c == nil {
		c = models.StringList{}
	}
	return map[string]any{
		"id": r.ID, "key": r.Key, "name": r.Name,
		"is_system": r.IsSystem, "is_superadmin": r.IsSuperadmin, "is_default": r.IsDefault,
		"capabilities": c,
	}
}

func filterCaps(in []string) models.StringList {
	out := models.StringList{}
	for _, c := range in {
		if caps.IsCap(c) {
			out = append(out, c)
		}
	}
	return out
}

// GET /roles
func (s *Server) listRoles(w http.ResponseWriter, r *http.Request) {
	var roles []models.Role
	s.db(r).Order("is_system DESC, id").Find(&roles)
	out := make([]map[string]any, 0, len(roles))
	for i := range roles {
		out = append(out, roleOut(&roles[i]))
	}
	writeJSON(w, http.StatusOK, out)
}

// POST /roles
func (s *Server) createRole(w http.ResponseWriter, r *http.Request) {
	var p struct {
		Key          string   `json:"key"`
		Name         string   `json:"name"`
		IsDefault    bool     `json:"is_default"`
		Capabilities []string `json:"capabilities"`
	}
	if err := decodeJSON(r, &p); err != nil {
		httpErr(w, http.StatusBadRequest, "Некорректный запрос")
		return
	}
	key := strings.ReplaceAll(strings.ToLower(strings.TrimSpace(p.Key)), " ", "_")
	name := strings.TrimSpace(p.Name)
	if name == "" {
		httpErr(w, http.StatusBadRequest, "Укажите название роли")
		return
	}
	if key == "" {
		key = strings.ReplaceAll(strings.ToLower(name), " ", "_")
	}
	var cnt int64
	s.db(r).Model(&models.Role{}).Where("key = ?", key).Count(&cnt)
	if cnt > 0 {
		httpErr(w, http.StatusBadRequest, "Роль с таким ключом уже существует")
		return
	}
	role := models.Role{
		Key: key, Name: name, IsSystem: false, IsSuperadmin: false,
		IsDefault: p.IsDefault, Capabilities: filterCaps(p.Capabilities),
	}
	if role.IsDefault {
		s.db(r).Model(&models.Role{}).Where("1 = 1").Update("is_default", false)
	}
	s.db(r).Create(&role)
	writeJSON(w, http.StatusCreated, roleOut(&role))
}

// PUT /roles/{id}
func (s *Server) updateRole(w http.ResponseWriter, r *http.Request) {
	id, _ := strconv.Atoi(chi.URLParam(r, "id"))
	var role models.Role
	if err := s.db(r).First(&role, id).Error; err != nil {
		httpErr(w, http.StatusNotFound, "Роль не найдена")
		return
	}
	if role.IsSuperadmin {
		httpErr(w, http.StatusBadRequest, "Роль гуру не редактируется")
		return
	}
	var p map[string]any
	if err := decodeJSON(r, &p); err != nil {
		httpErr(w, http.StatusBadRequest, "Некорректный запрос")
		return
	}
	upd := map[string]any{}
	if v, ok := p["name"].(string); ok && strings.TrimSpace(v) != "" {
		upd["name"] = strings.TrimSpace(v)
	}
	if raw, ok := p["capabilities"].([]any); ok {
		list := make([]string, 0, len(raw))
		for _, x := range raw {
			if s, ok := x.(string); ok {
				list = append(list, s)
			}
		}
		upd["capabilities"] = filterCaps(list)
	}
	if v, ok := p["is_default"]; ok {
		def, _ := v.(bool)
		upd["is_default"] = def
		if def {
			s.db(r).Model(&models.Role{}).Where("id <> ?", role.ID).Update("is_default", false)
		}
	}
	if len(upd) > 0 {
		s.db(r).Model(&models.Role{}).Where("id = ?", role.ID).Updates(upd)
		s.db(r).First(&role, role.ID)
	}
	writeJSON(w, http.StatusOK, roleOut(&role))
}

// DELETE /roles/{id}
func (s *Server) deleteRole(w http.ResponseWriter, r *http.Request) {
	id, _ := strconv.Atoi(chi.URLParam(r, "id"))
	var role models.Role
	if err := s.db(r).First(&role, id).Error; err != nil {
		httpErr(w, http.StatusNotFound, "Роль не найдена")
		return
	}
	if role.IsSystem {
		httpErr(w, http.StatusBadRequest, "Системную роль удалить нельзя")
		return
	}
	s.DB.Where("role_id = ?", role.ID).Delete(&models.UserRole{})
	s.db(r).Delete(&models.Role{}, role.ID)
	w.WriteHeader(http.StatusNoContent)
}

// GET /users/{id}/roles
func (s *Server) getUserRoles(w http.ResponseWriter, r *http.Request) {
	id, _ := strconv.Atoi(chi.URLParam(r, "id"))
	var urs []models.UserRole
	s.DB.Where("user_id = ?", id).Find(&urs)
	ids := make([]int, 0, len(urs))
	for _, ur := range urs {
		ids = append(ids, ur.RoleID)
	}
	writeJSON(w, http.StatusOK, map[string]any{"role_ids": ids})
}

// PUT /users/{id}/roles
func (s *Server) setUserRoles(w http.ResponseWriter, r *http.Request) {
	id, _ := strconv.Atoi(chi.URLParam(r, "id"))
	var u models.User
	if err := s.DB.First(&u, id).Error; err != nil {
		httpErr(w, http.StatusNotFound, "Пользователь не найден")
		return
	}
	var p struct {
		RoleIDs []int `json:"role_ids"`
	}
	if err := decodeJSON(r, &p); err != nil {
		httpErr(w, http.StatusBadRequest, "Некорректный запрос")
		return
	}
	valid := []int{}
	if len(p.RoleIDs) > 0 {
		var roles []models.Role
		s.db(r).Where("id IN ?", p.RoleIDs).Find(&roles)
		for _, r := range roles {
			valid = append(valid, r.ID)
		}
	}
	s.DB.Where("user_id = ?", id).Delete(&models.UserRole{})
	for _, rid := range valid {
		s.DB.Create(&models.UserRole{UserID: id, RoleID: rid})
	}
	sort.Ints(valid)
	writeJSON(w, http.StatusOK, map[string]any{"role_ids": valid})
}
